man/man1/rasort.1
.\" Copyright (c) 2000-2008 QoSient, LLC
.\" All rights reserved.
.\" 
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2, or (at your option)
.\" any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH RASORT 1 "07 November 2000"
.SH NAME
\fBrasort\fP \- sort \fBargus(8)\fP data file.
.SH COPYRIGHT
Copyright (c) 2000-2008 QoSient. All rights reserved.
.SH SYNOPSIS
.B rasort
[[\fB\-M\fP \fIsortmode\fP] [\fB\-m\fP \fIsortmode\fP] ...]
[\fBraoptions\fP]
.SH DESCRIPTION
.IX  "rasort command"  ""  "\fLrasort\fP \(em argus data"
.LP
.B Rasort
reads
.BR argus
data from an \fIargus-data\fP source, sorts the records based on
the criteria specified on the command line, and outputs a valid
\fIargus-stream\fP.
.SH OPTIONS
Rasort, like all ra based clients, supports a number of
\fBra options\fP including filtering of input argus
records through a terminating filter expression.
See \fBra(1)\fP for a complete description of \fBra options\fP.
\fBrasort(1)\fP specific options are:
.PP
.PD 0
.TP 4 4
.B \-M replace
Replace the existing file with the sorted output.

.TP 15
.BI \-m "\| sortmode\^"
Supported sortmodes are:
.PP
.RS
.TP 15
.B stime
record start time <default>
.TP
.B ltime
record last time.
.TP
.B trans
aggregation record count.
.TP
.B dur
record total duration.
.TP
.B avgdur
record average duration.
.TP
.B mindur
record minimum duration.
.TP
.B maxdur
record maximum duration.
.TP
.B smac
source MAC addr.
.TP
.B dmac
destination MAC addr.
.TP
.B saddr
source IP addr.
.TP
.B daddr
destination IP addr.
.TP
.B proto
transaction protocol.
.TP
.B sport
source port number.
.TP
.B dport
destination port number.
.TP
.B stos
source TOS byte value.
.TP
.B dtos
destination TOS byte value.
.TP
.B sttl
src -> dst TTL value.
.TP
.B dttl
dst -> src TTL value.
.TP
.B bytes
total transaction bytes.
.TP
.B sbytes
src -> dst transaction bytes.
.TP
.B dbytes
dst -> src transaction bytes.
.TP
.B pkts
total transaction packet count.
.TP
.B spkts
src -> dst packet count.
.TP
.B dpkts
dst -> src packet count.
.TP
.B load
bits per second.
.TP
.B sload
source bits per second.
.TP
.B dload
destination bits per second.
.TP
.B loss
pkts retransmitted or dropped.
.TP
.B sloss
source pkts retransmitted or dropped.
.TP
.B dloss
destination pkts retransmitted or dropped.
.TP
.B ploss
percent pkts retransmitted or dropped.
.TP 
.B sploss
percent source pkts retransmitted or dropped. 
.TP
.B dploss
percent destination pkts retransmitted or dropped.
.TP
.B rate
pkts per second.
.TP
.B srate
source pkts per second.
.TP
.B drate
destination pkts per second.
.TP
.B tranref
argus transaction reference number.
.TP
.B seq
argus sequence number.
.TP
.B smpls
source MPLS identifier.
.TP
.B dmpls
destination MPLS identifier.
.TP
.B svlan
source VLAN identifier.
.TP
.B dvlan
destination VLAN identifier.
.TP
.B srcid
argus source identifier.
.TP
.B stcpb
source TCP base sequence number.
.TP
.B dtcpb
destination TCP base sequence number.
.TP
.B tcprtt
TCP connection setup round-trip time.
.PD
.RE

.SH INVOCATION
A sample invocation of \fBrasort(1)\fP.  This call reads \fBargus(8)\fP data
from \fBinputfile\fP and sorts the IP protocol based \fBargus(8)\fP data, first
by the destination IP address, then by the service (destination) port number
and then by the source IP address, and writes the results to stdout.
For most services, this arranges \fBargus(8)\fP formatted data by server,
service, and then by client.
.TP 5
\fBrasort\fP -r inputfile -m daddr dport saddr - ip
.SH SEE ALSO
.BR ra(1),
.BR rarc(5),
.BR argus(8),
.SH FILES

.SH AUTHORS
.nf
Carter Bullard (carter@qosient.com).
.fi
.SH BUGS


ChangeLog
Fri Jul 31 11:47:29 EDT 2009
Fix for 2.0 record conversion.  rtcp dsr inserted with len of 1 and
ArgusNtoH was corrupting other DSRs trying to little-endian the field.

Wed Jul 29 13:46:38 EDT 2009
Wow, "-S 1.2.3.4" didn't work, so fixed that.

Wed Jul 22 00:46:11 EDT 2009
Fix for trans processing in rahisto().
Modification of 'autoid' support in rasqlinsert().

Thu Jul  9 15:35:05 EDT 2009
Fix for rasplit() bug where the $srcid directive is ignored in some
situations. Not good.

Wed Jul  8 15:49:13 EDT 2009
Fix for TCP synAck flows.  Extend appbyte check in case argus wasn't 
configured to generate appbytes.
Fix for ratop ^L key.

Fri Jun 19 12:20:24 EDT 2009
Fix for ASN for inode IP address little endian converstion problem.

Mon Jun 15 09:12:25 EDT 2009
Major fix for processing IPv6 ICMP data.
Major fix for processing User data DSR.
Fix to not print field delimiter when  printing fields.

Thu May 21 13:11:59 EDT 2009
New fix for rasplit when spliting on anything bigger than hours.
New faster strategy for dealing with daylight savings time issues.
Updates for rabins for new spliting strategies.

Fri May 15 13:26:09 EDT 2009
Mods to fix RA_FIELD_WIDTH behavior.

Thu May  7 12:47:06 EDT 2009
Massive changes for Geolocation.
AS number reporting (change as from short to int).
Improved DNS reporting formats so field number is constant.
Fixed merging record problem.

Mon Apr 13 17:55:47 EDT 2009
Modified RA_FIELD_WIDTH behavior with using '-c' on the command line.
Added RA_FIELD_WIDTH to rarc.

Fixed problem with RaPrintIndex and rasqlinsert.
Modified rasqlinsert() minimum field width to be 256, so that addresses
will always be big enough.

Modified RA_TIME_FORMAT string parsing ala Dan Klien, to get non time
oriented characters and escaped sequences parsed correctly.

Mon Apr  6 11:37:36 EDT 2009
Fixed v2 -> v3 conversion.  ArgusNetworkStruct len was a few bytes too long.
Fixed mode parsing for rasplit().
Fix for rasplit() finding filename on split 'd','w','m','y'.  Algorithm
chose GMT not local timezone.

Fri Apr  3 12:53:15 EDT 2009
Fixes for rasqlinsert() to better deal with strftime() based table names.
Fixed ratree() to work in argus-3.0.

Mon Mar 30 16:34:24 EDT 2009
Added table date specifications for rasqlinsert().
Fix for memory problem when opening lots of files or directories.
Added RA_TIMEOUT_INTERVAL to rarc file.
Fixed short DSR parsing.

Fri Mar 27 09:41:05 EDT 2009
Fix for tcprtt propagation during merge when first record
is a ARGUS_TCP_INIT type DSR.

Fix for little-endian treatment of Authoritative response
in DNS printing routines of radump().

Wed Mar 25 16:09:53 EDT 2009
Mods to complete library to fix issues with -M rmon option.
Problem caused duplicate records generated by "-M rmon"
to be corrected before being aggregated.  Big problem,
basically just doubling metrics, but not flow ids.

Mon Mar 23 14:51:21 EDT 2009
Mod to ArgusClientInit in rabins() to deal with calling without
a -B option to specify the range.  Seems to work now.

Fri Mar 13 14:53:46 EDT 2009
Minor changes for raservices and rauserdata.
May have broken "finding" possible protocol types, need
to work on this this  weekend.

Thu Mar 12 11:40:14 EDT 2009
Removed -lz from aclocal.m4 an configure.in.  Will need
to add it back soon.

Wed Mar 11 20:30:17 EDT 2009
Fixed record searching when the read buffer boundary is
too short for searching for new data record.  Reordered
logic to do test, when enough bytes are available, and
to break way out if not enough bytes.

Mon Mar  9 12:45:03 EDT 2009
Fixed aggregation rules for FRAGMENT reporting in flow.
argus changes its reporting of FRAGMENTS scope to the single
flow record, and that revealed that merging was messing
up the IP attributes flags.  

Fri Mar  6 13:38:36 EST 2009
Added record searching for argus-2.x data sets when out of sync.
fixed early file closing for ra*.

Mon Mar  2 14:28:02 EST 2009
Release candidate #1 for clients-3.0.2.  This has all the support
for next wave argus DSRs (histograms and events), and many bug
fixes from the mailing list and beyond.

Mods to TCP direction corrections, to be more strict on what
we'll reverse.

Fixes for backward compatiblity for argus-2.x data streams.

Added rauserdata and rasservices to distribution to assist in user
data analysis.

Wed Feb 25 16:01:36 EST 2009
Complete integration of gargoyle client code back to argus,
without IB and control plane flow monitoring.  Major
rewrite of rabins(), ratop() and racluster() for ARMA
functions.  Completely differnt compiler filtering strategies
to allow for DSR specific filtering.

Addition of mysql insertion and query programs, but no NTAIS
project support.

Elimination of ragrep(), by merging grep capabilities into all
ra* programs.

Tue Feb 24 10:07:24 EST 2009
Major mods to fix graphing problems exposed by ArgusGraph and Topology.
Added sub-second bins, and better reliablity.

Tue Oct 28 23:32:29 EDT 2008
Major modifications for introducing new linear histogram
reporting for interpacket arrival times.

Mon Oct 27 10:44:42 EDT 2008
Major modifications for restructuring library strategies to accomodate
new client architectures.

Fri Sep  5 13:27:01 EDT 2008
Fixes for rasql, rasqlinsert, all is well.
Terminatd xml ArgusDataStream in rasql.

Fri Aug  1 10:39:35 EDT 2008
Massive changes for all things.
MySQL programs now working well.

Tue Apr  8 14:55:09 EDT 2008
Fixes for time filter problem and addition of UTC input.
Fixes for radump() to parse user buffer.

Sat Mar 15 17:14:31 EDT 2008
Fixes for printing when srcid is not an IP address.

Thu Feb 28 19:38:40 EST 2008
Fixes for rastream() and to get the right readline
on the mac.

Wed Feb 13 00:17:15 EST 2008
Added support for Multipath

Mon Feb 11 22:45:40 EST 2008
Mods for putting listen() down on all address families.
Mod for Makefiles to remove .PHONY for OpenBSD

Tue Jan 29 13:36:49 EST 2008
Fixes for new flex 2.5.34 issues

Mon Jan 21 20:03:26 EST 2008
Fixes for default unidirectional flow record matching

clients/rasort.c
/*
 * Argus Software
 * Copyright (c) 2000-2009 QoSient, LLC
 * All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.

 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * 
 * $Id: //depot/argus/clients/clients/rasort.c#26 $
 * $DateTime: 2009/05/15 12:40:06 $
 * $Change: 1732 $
 */

/*
 * rasort.c  - sort argus records based on various fields.
 *       
 * written by Carter Bullard
 * QoSient, LLC
 *
 */

#if defined(CYGWIN)
#define USE_IPV6
#endif

#include <unistd.h>
#include <stdlib.h>

#include <compat.h>

#include <rabins.h>
#include <argus_util.h>
#include <argus_client.h>
#include <argus_main.h>
#include <argus_sort.h>
#include <argus_filter.h>

#include <signal.h>
#include <ctype.h>


void
ArgusClientInit (struct ArgusParserStruct *parser)
{
   struct ArgusModeStruct *mode;
   int i = 0, x = 0;

   if (!(parser->RaInitialized)) {
      (void) signal (SIGHUP,  (void (*)(int)) RaParseComplete);
      (void) signal (SIGTERM, (void (*)(int)) RaParseComplete);
      (void) signal (SIGQUIT, (void (*)(int)) RaParseComplete);
      (void) signal (SIGINT,  (void (*)(int)) RaParseComplete);

      parser->RaWriteOut = 0;

      if (parser->vflag)
         ArgusReverseSortDir++;

      if ((ArgusSorter = ArgusNewSorter()) == NULL)
         ArgusLog (LOG_ERR, "ArgusClientInit ArgusNewSorter error %s", strerror(errno));

      bzero ((char *) ArgusSorter->ArgusSortAlgorithms, sizeof(ArgusSorter->ArgusSortAlgorithms));
      ArgusSorter->ArgusSortAlgorithms[0] = ArgusSortAlgorithmTable[0];
 
      if ((mode = parser->ArgusModeList) != NULL) {
         while (mode) {
            if (!(strcmp ("replace", mode->mode))) {
               ArgusSorter->ArgusReplaceMode++;
               if ((parser->ArgusWfileList != NULL) && (!(ArgusListEmpty(parser->ArgusWfileList)))) {
                  ArgusLog (LOG_ERR, "replace mode and -w option are incompatible\n");
               }
            }
 
            mode = mode->nxt;
         }
      }

      if ((mode = parser->ArgusMaskList) != NULL) {
         while (mode) {
            for (x = 0; x < MAX_SORT_ALG_TYPES; x++) {
               if (!strncmp (ArgusSortKeyWords[x], mode->mode, strlen(ArgusSortKeyWords[x]))) {
                  ArgusSorter->ArgusSortAlgorithms[i++] = ArgusSortAlgorithmTable[x];
                  break;
               }
            }

            if (x == MAX_SORT_ALG_TYPES)
               ArgusLog (LOG_ERR, "sort syntax error. \'%s\' not supported", mode->mode);

            mode = mode->nxt;
         }
      }

      parser->RaInitialized++;
   }
}

void
RaArgusInputComplete (struct ArgusInput *input)
{
   struct ArgusRecordStruct *nsr;
   struct ArgusWfileStruct *wfile = NULL;
   char buf[MAXSTRLEN];
   int count, label, i, fd;
 
   if (ArgusSorter->ArgusReplaceMode) {
      if (ArgusParser->ArgusWfileList == NULL)
         ArgusParser->ArgusWfileList = ArgusNewList();
 
      if ((count = ArgusSorter->ArgusRecordQueue->count) > 0) {
         if (!(ArgusParser->ArgusRandomSeed))
            srandom(ArgusParser->ArgusRandomSeed);

         srandom (ArgusParser->ArgusRealTime.tv_usec);
         label = random() % 100000;
 
         bzero(buf, sizeof(buf));
         snprintf (buf, MAXSTRLEN, "%s.tmp%d", input->filename, label);
         if ((fd = open(buf, O_CREAT|O_EXCL, input->statbuf.st_mode)) < 0)
            ArgusLog (LOG_ERR, "open %s error: %s", buf, strerror(errno));
 
         close(fd);
 
         if ((wfile = (struct ArgusWfileStruct *) ArgusCalloc (1, sizeof (*wfile))) != NULL) {
            ArgusPushFrontList(ArgusParser->ArgusWfileList, (struct ArgusListRecord *)wfile, ARGUS_NOLOCK);
            wfile->filename  = strdup(buf);
 
         } else
            ArgusLog (LOG_ERR, "setArgusWfile, ArgusCalloc %s", strerror(errno));
 
         ArgusSortQueue (ArgusSorter, ArgusSorter->ArgusRecordQueue);
 
         for (i = 0, count = ArgusSorter->ArgusRecordQueue->count; i < count; i++)
            RaSendArgusRecord ((struct ArgusRecordStruct *)ArgusSorter->ArgusRecordQueue->array[i]);
 
         while ((nsr = (struct ArgusRecordStruct *) ArgusPopQueue(ArgusSorter->ArgusRecordQueue, ARGUS_NOLOCK)) != NULL)
            ArgusDeleteRecordStruct(ArgusParser, nsr);
 
         rename (wfile->filename, input->filename);
         fclose (wfile->fd);
         ArgusDeleteList (ArgusParser->ArgusWfileList, ARGUS_WFILE_LIST);
         ArgusParser->ArgusWfileList = NULL;
 
         if (ArgusParser->Vflag)
            ArgusLog(LOG_INFO, "file %s sorted", input->filename);
      }
   }
}


void
RaParseComplete (int sig)
{
   int i = 0, count = 0;
 
   if (sig >= 0) {
      if (!ArgusParser->RaParseCompleting++) {
         count = ArgusSorter->ArgusRecordQueue->count;

         if (count > 0) {
            ArgusSortQueue (ArgusSorter, ArgusSorter->ArgusRecordQueue);
 
            for (i = 0; i < count; i++)
               RaSendArgusRecord ((struct ArgusRecordStruct *) ArgusSorter->ArgusRecordQueue->array[i]);
         }
      }

#ifdef ARGUSDEBUG
      ArgusDebug (2, "RaParseComplete(caught signal %d)\n", sig);
#endif
      switch (sig) {
         case SIGHUP:
         case SIGINT:
         case SIGTERM:
         case SIGQUIT: {
            struct ArgusWfileStruct *wfile = NULL;

            if (ArgusParser->ArgusWfileList != NULL) {
               struct ArgusListObjectStruct *lobj = NULL;
               int i, count = ArgusParser->ArgusWfileList->count;

               if ((lobj = ArgusParser->ArgusWfileList->start) != NULL) {
                  for (i = 0; i < count; i++) {
                     if ((wfile = (struct ArgusWfileStruct *) lobj) != NULL) {
                        if (wfile->fd != NULL) {
#ifdef ARGUSDEBUG
                           ArgusDebug (2, "RaParseComplete: closing %s\n", wfile->filename);
#endif
                           fflush (wfile->fd);
                           fclose (wfile->fd);
                           wfile->fd = NULL;
                        }
                     }
                     lobj = lobj->nxt;
                  }
               }
            }
            exit(0);
            break;
         }
      }
   }
}


void
ArgusClientTimeout ()
{
#ifdef ARGUSDEBUG
   ArgusDebug (6, "ArgusClientTimeout()\n");
#endif
}

void
parse_arg (int argc, char**argv)
{}

void
usage ()
{
   extern char version[]; 
             
   fprintf (stderr, "Rasort Version %s\n", version);
   fprintf (stderr, "usage: %s [[-M mode] [-m sortfield] ...] [ra-options] [- filter-expression]\n", ArgusParser->ArgusProgramName);
   fprintf (stderr, "options:    -M replace         replace the original file with the sorted output.\n");
   fprintf (stderr, "            -m <sortfield>     specify the <sortfield>(s) in order.\n");
   fprintf (stderr, "                               valid sorfields are:\n");
   fprintf (stderr, "                                  stime, ltime, trans, dur, avgdur, mindur,\n");
   fprintf (stderr, "                                  maxdur, smac, dmac, saddr, daddr, proto, \n");
   fprintf (stderr, "                                  sport, dport, stos, dtos, sttl, dttl, bytes,\n");
   fprintf (stderr, "                                  sbytes, dbytes, pkts, spkts, dpkts, load,\n"); 
   fprintf (stderr, "                                  sload, sload, dload, loss, sloss, dloss,\n"); 
   fprintf (stderr, "                                  ploss, psloss, pdloss, rate, srate, drate,\n"); 
   fprintf (stderr, "                                  seq, smpls, dmpls, svlan, dvlan, srcid,\n");
   fprintf (stderr, "                                  stcpb, dtcpb, tcprtt\n"); 
   fprintf (stderr, "\n"); 
   fprintf (stderr, "ra-options: -b                 dump packet-matching code.\n");
   fprintf (stderr, "            -C <[host]:<port>  specify remote Cisco Netflow source.\n");
   fprintf (stderr, "                               source.\n");
#if defined (ARGUSDEBUG)
   fprintf (stderr, "            -D <level>         specify debug level\n");
#endif
   fprintf (stderr, "            -F <conffile>      read configuration from <conffile>.\n");
   fprintf (stderr, "            -h                 print help.\n");
   fprintf (stderr, "            -r <file>          read argus data <file>. '-' denotes stdin.\n");
   fprintf (stderr, "            -s [-][+[#]]field  specify fields to print.\n");
   fprintf (stderr, "            -S <host[:port]>   specify remote argus <host> and optional port\n");
   fprintf (stderr, "                               number.\n");
   fprintf (stderr, "            -t <timerange>     specify <timerange> for reading records.\n");
   fprintf (stderr, "                      format:  timeSpecification[-timeSpecification]\n");
   fprintf (stderr, "                               timeSpecification: [mm/dd[/yy].]hh[:mm[:ss]]\n");
   fprintf (stderr, "                                                   mm/dd[/yy]\n");
   fprintf (stderr, "                                                   -%%d{yMhdms}\n");
   fprintf (stderr, "            -T <secs>          attach to remote server for T seconds.\n");
#ifdef ARGUS_SASL
   fprintf (stderr, "            -U <user/auth>     specify <user/auth> authentication information.\n");
#endif
   fprintf (stderr, "            -w <file>          write output to <file>. '-' denotes stdout.\n");
   exit(1);
}

void
RaProcessRecord (struct ArgusParserStruct *parser, struct ArgusRecordStruct *ns)
{
   struct ArgusRecordStruct *tns = NULL;

   switch (ns->hdr.type & 0xF0) {
      case ARGUS_MAR:
      case ARGUS_EVENT: {
         break;
      }

      case ARGUS_NETFLOW:
      case ARGUS_FAR: {
         if ((tns = ArgusCopyRecordStruct(ns)) == NULL)
            ArgusLog (LOG_ERR, "RaProcessRecord: ArgusCopyRecordStruct(0x%x) error\n", ns);

         ArgusAddToQueue (ArgusSorter->ArgusRecordQueue, &tns->qhdr, ARGUS_NOLOCK);
         break;
      }
   }
}

int
RaSendArgusRecord(struct ArgusRecordStruct *ns)
{
   int retn = 1;
   char buf[MAXSTRLEN];

   if (ns->status & ARGUS_RECORD_WRITTEN)
      return (retn);
 
   if ((ArgusParser->ArgusWfileList != NULL) && (!(ArgusListEmpty(ArgusParser->ArgusWfileList)))) {
      struct ArgusWfileStruct *wfile = NULL;
      struct ArgusListObjectStruct *lobj = NULL;
      int i, count = ArgusParser->ArgusWfileList->count;

      if ((lobj = ArgusParser->ArgusWfileList->start) != NULL) {
         for (i = 0; i < count; i++) {
            if ((wfile = (struct ArgusWfileStruct *) lobj) != NULL) {
               int pass = 1;
               if (wfile->filterstr) {
                  struct nff_insn *wfcode = wfile->filter.bf_insns;
                  pass = ArgusFilterRecord (wfcode, ns);
               }

               if (pass != 0) {
                  if ((ArgusParser->exceptfile == NULL) || strcmp(wfile->filename, ArgusParser->exceptfile)) {
                     struct ArgusRecord *argusrec = NULL;
                     char buf[2048];
                     if ((argusrec = ArgusGenerateRecord (ns, 0L, buf)) != NULL) {
#ifdef _LITTLE_ENDIAN
                        ArgusHtoN(argusrec);
#endif
                        ArgusWriteNewLogfile (ArgusParser, ns->input, wfile, argusrec);
                     }
                  }
               }
            }
            lobj = lobj->nxt;
         }
      }

   } else {
      if (!ArgusParser->qflag) {
         if (ArgusParser->Lflag) {
            if (ArgusParser->RaLabel == NULL)
               ArgusParser->RaLabel = ArgusGenerateLabel(ArgusParser, ns);
 
            if (!(ArgusParser->RaLabelCounter++ % ArgusParser->Lflag))
               printf ("%s\n", ArgusParser->RaLabel);
 
            if (ArgusParser->Lflag < 0)
               ArgusParser->Lflag = 0;
         }

         *(int *)&buf = 0;
         ArgusPrintRecord(ArgusParser, buf, ns, MAXSTRLEN);
         fprintf (stdout, "%s\n", buf);
         fflush(stdout);
      }
   }

   ns->status |= ARGUS_RECORD_WRITTEN;
   return (retn);
}

void ArgusWindowClose(void);

void ArgusWindowClose(void) { 
#ifdef ARGUSDEBUG
   ArgusDebug (6, "ArgusWindowClose () returning\n"); 
#endif
}



